src/deploy/functions/validate.ts

```diff
@@ -49,3 +53,73 @@ export function functionIdsAreValid(functions: { id: string; platform: string }[
     throw new FirebaseError(msg);
   }
 }
+
+/**
+ * Validate secret environment variables setting, if any.
+ * A bad secret configuration can lead to a significant delay in function deploys.
+ *
+ * If validation fails for any secret config, throws a FirebaseError.
+ */
+export async function secretsAreValid(b: backend.Backend) {
+  const endpoints = backend
+    .allEndpoints(b)
+    .filter((e) => e.secretEnvironmentVariables && e.secretEnvironmentVariables.length > 0);
+  validatePlatformTargets(endpoints);
+  await validateSecretVersions(endpoints);
+}
+
+/**
+ * Ensures that all endpoints specifying secret environment variables target platform that supports the feature.
+ */
+function validatePlatformTargets(endpoints: backend.Endpoint[]) {
+  const supportedPlatforms = ["gcfv1"];
+  const unsupported = endpoints.filter((e) => !supportedPlatforms.includes(e.platform));
+  if (unsupported.length > 0) {
+    const errs = unsupported.map((e) => `${e.id}[platform=${e.platform}]`);
+    throw new FirebaseError(
+      `Tried to set secret environment variables on ${errs.join(", ")}. ` +
+        `Only ${supportedPlatforms.join(", ")} support secret environments.`
+    );
+  }
+}
+
+/**
+ * Validate each secret version referenced in target endpoints.
+ *
+ * A secret version is valid if:
+ *   1) It exists.
+ *   2) It's in state "enabled".
+ */
+async function validateSecretVersions(endpoints: backend.Endpoint[]) {
+  const validate = async (s: backend.SecretEnvVar) => {
+    const sv = await getSecretVersion(s.projectId, s.secret, "latest");
```

Comment:
What's the permission level required for this API call? Is it something we should do to assist customers when we have permissions but otherwise be silent?

Reply:
I don't think permission issues raised here is going to be confusing - you either have permission to list secrets in a project or you don't. 

Ensuring secret access code - that I think will benefit from some hand-holding. Will seek change there.